posthog/clickhouse/client/async_task_chain.py

```diff
@@ -0,0 +1,87 @@
+import threading
+import typing
+from contextlib import contextmanager
+from functools import partial
+
+from celery import chain
+from celery.canvas import Signature
+from django.db import transaction
+
+from posthog.schema import QueryStatus
+
+if typing.TYPE_CHECKING:
+    from posthog.clickhouse.client.execute_async import QueryStatusManager
+
+# Create a thread-local storage for task chains
+_thread_locals = threading.local()
+
+
+def kick_off_task(
+    manager: "QueryStatusManager",
+    query_status: QueryStatus,
+    task_signature: Signature,
+) -> None:
+    task = task_signature.apply_async()
+    query_status.task_id = task.id
+    manager.store_query_status(query_status)
+
+
+def get_task_chain() -> list[tuple[Signature, "QueryStatusManager", QueryStatus]]:
+    """
+    Retrieves the task chain from thread-local storage.
+    """
+    if not hasattr(_thread_locals, "task_chain"):
+        _thread_locals.task_chain = []
+    return _thread_locals.task_chain
+
+
+def set_in_context(value) -> None:
+    """
+    Sets the in_context flag in thread-local storage.
+    """
+    _thread_locals.in_context = value
+
+
+def is_in_context() -> bool:
+    """
+    Checks if the in_context flag is set in thread-local storage.
+    """
+    return getattr(_thread_locals, "in_context", False)
+
+
+def add_task_to_chain(task_signature: Signature, manager: "QueryStatusManager", query_status: QueryStatus) -> None:
+    """
```

Comment:
The ability to use `add_task_to_chain()` without an actual chain context feels like a bit of a footgun. It's not really `add_task_to_chain()` then, just `add_task_to_on_commit()`. Would it make sense to throw here if not in context, for clarity?